src/language-server/project/rover/project.ts

```diff
@@ -50,56 +67,93 @@ export class RoverProject extends GraphQLProject {
+  /**
+   * Since Rover projects do not scan all the folders in the workspace on start,
+   * we need to restore the information about open documents from the previous session
+   * in case or a project recreation (configuration reload).
+   */
+  restoreFromPreviousProject(previousProject: RoverProject) {
+    for (const document of previousProject.documents.openDocuments) {
+      this.documents.onDidOpenTextDocument({ document });
     }
-    return (this._connection = this.initializeConnection());
+  }
+
+  getConnection(): Promise<ProtocolConnection> {
+    const connectionFromStorage = this.connectionStorage.getStore();
+    if (connectionFromStorage) {
+      return Promise.resolve(connectionFromStorage);
+    }
+
+    if (!this._connection) {
+      this._connection = this.initializeConnection();
+    }
+
+    return this._connection;
   }
 
   private async sendNotification<P, RO>(
     type: ProtocolNotificationType<P, RO>,
     params?: P,
   ): Promise<void> {
-    const connection = await this.connection;
-    console.log("sending notification", { type, params });
-    return connection.sendNotification(type, params);
+    const connection = await this.getConnection();
+    DEBUG &&
+      console.log("sending notification %o", {
+        type: type.method,
+        params,
+      });
+    try {
+      return await connection.sendNotification(type, params);
+    } catch (error) {
+      if (error instanceof Error) {
+        (error as any).cause = { type: type.method, params };
+      }
+      throw error;
+    }
   }
 
   private async sendRequest<P, R, PR, E, RO>(
     type: ProtocolRequestType<P, R, PR, E, RO>,
     params: P,
     token?: CancellationToken,
   ): Promise<R> {
-    const connection = await this.connection;
-    console.log("sending request", { type, params });
-    return connection
-      .sendRequest(type, params, token)
-      .then((result) => {
-        console.log({ result });
-        return result;
-      })
-      .catch((error) => {
-        console.error({ error });
-        throw error;
-      });
+    const connection = await this.getConnection();
+    DEBUG && console.log("sending request %o", { type: type.method, params });
+    try {
+      const result = await connection.sendRequest(type, params, token);
+      DEBUG && console.log({ result });
+      return result;
+    } catch (error) {
+      if (error instanceof Error) {
+        (error as any).cause = { type: type.method, params };
+      }
+      throw error;
+    }
   }
 
-  initializeConnection() {
+  async initializeConnection() {
+    if (this.child) {
+      this.child.kill();
+    }
+    if (this.disposed) {
+      throw new ConnectionError(
+        ConnectionErrors.Closed,
+        "Connection is closed.",
+      );
+    }
     const child = cp.spawn(this.config.rover.bin, ["lsp"], {
-      env: { RUST_BACKTRACE: "1" },
+      env: DEBUG ? { RUST_BACKTRACE: "1" } : {},
+      stdio: ["pipe", "pipe", DEBUG ? "inherit" : "ignore"],
```

Comment:
`"inherit"` here would just re-print STDERR to the parent process STDERR - it replaces 

```
    child.stderr.on("data", (data) => {
      console.info("stderr", data.toString());
    });
```